self-paced-labs/vertex-ai/vertex-ai-qwikstart/lab_exercise.ipynb

```diff
@@ -1567,158 +1717,384 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1456,
+   "id": "6aa86895",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Overwriting online-retail-clv-3M/Dockerfile\n"
+     ]
+    }
+   ],
+   "source": [
+    "%%writefile {MODEL_NAME}/Dockerfile\n",
+    "# Specifies base image and tag.\n",
+    "# https://cloud.google.com/vertex-ai/docs/training/pre-built-containers\n",
+    "# us-docker.pkg.dev/vertex-ai/training/tf-cpu.2-3:latest\n",
+    "FROM gcr.io/deeplearning-platform-release/tf2-cpu.2-3\n",
     "\n",
-    "model.fit(training_ds,\n",
-    "          validation_data=eval_ds,\n",
-    "          steps_per_epoch=STEPS_PER_EPOCH,\n",
-    "          epochs=N_CHECKPOINTS,\n",
-    "          callbacks=[[tensorboard_callback,\n",
-    "                      earlystopping_callback,\n",
-    "                      checkpoint_callback]])\n",
+    "# Sets the container working directory.\n",
+    "WORKDIR /root\n",
     "\n",
-    "print(model.evaluate(test_ds))\n",
+    "# Copies the requirements.txt into the container to reduce network calls.\n",
+    "COPY requirements.txt .\n",
+    "# Installs additional packages.\n",
+    "RUN pip3 install -U -r requirements.txt\n",
     "\n",
-    "tmpdir = tempfile.mkdtemp()\n",
+    "# Copies the trainer code to the docker image.\n",
+    "COPY . /trainer\n",
     "\n",
-    "# Export Keras model in TensorFlow SavedModel format.\n",
-    "model.save(tmpdir)\n",
+    "# Sets the container working directory.\n",
+    "WORKDIR /trainer\n",
     "\n",
-    "# Annotate and save TensorFlow SavedModel with Explainable metadata to GCS.\n",
-    "builder = SavedModelMetadataBuilder(tmpdir)\n",
-    "builder.save_model_with_metadata(os.environ[\"AIP_MODEL_DIR\"])"
+    "# Sets up the entry point to invoke the trainer.\n",
+    "ENTRYPOINT [\"python\", \"-m\", \"trainer.task\"]"
    ]
   },
   {
    "cell_type": "markdown",
-   "id": "cb3ac018",
+   "id": "90d9fecd",
    "metadata": {},
    "source": [
-    "### Use Cloud Build to build and submit container to Google Cloud Container Registry"
+    "### 4. Write a `requirements.txt` file to specify additional ML code dependencies"
    ]
   },
   {
    "cell_type": "markdown",
-   "id": "71d8344f",
+   "id": "279d289c",
    "metadata": {},
    "source": [
-    "Next, you will use [Cloud Build](https://cloud.google.com/build) to build and upload your custom TensorFlow model container to [Google Cloud Container Registry](https://cloud.google.com/container-registry).\n",
+    "These are additional dependencies for your model code outside the pre-built containers needed for prediction explainability and the BigQuery TensorFlow IO reader."
```

Comment:
It is possible to supply Python dependencies to pre-built containers too.

So either we need to move to true pre-built containers, or identify another reason why we need custom.

https://cloud.google.com/vertex-ai/docs/training/create-python-pre-built-container#working_with_dependencies

Reply:
Updated explanation. Packaging ML model code + dependencies in single container brings portability, stronger reproducibility guarantees, and onboarding workflow to Vertex Pipelines.